File: cliffy-protocols/src/crdt.rs
//! Geometric CRDT implementations using Clifford algebra

use crate::serde_ga3;
use crate::vector_clock::VectorClock;
use cliffy_core::GA3;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

/// A CRDT that uses geometric algebra operations for conflict resolution
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeometricCRDT {
    #[serde(with = "serde_ga3")]
    pub state: GA3,
    pub vector_clock: VectorClock,
    pub node_id: Uuid,
    pub operations: HashMap<u64, GeometricOperation>,
}

/// A geometric operation that can be applied to the CRDT state
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeometricOperation {
    pub id: u64,
    pub node_id: Uuid,
    pub timestamp: VectorClock,
    #[serde(with = "serde_ga3")]
    pub transform: GA3,
    pub operation_type: OperationType,
}

/// Types of geometric operations supported by the CRDT
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum OperationType {
    GeometricProduct,
    Addition,
    Sandwich,
    Exponential,
}

impl GeometricCRDT {
    /// Create a new GeometricCRDT with the given initial state
    pub fn new(node_id: Uuid, initial_state: GA3) -> Self {
        Self {
            state: initial_state,
            vector_clock: VectorClock::new(),
            node_id,
            operations: HashMap::new(),
        }
    }

    /// Apply a geometric operation to the CRDT state
    pub fn apply_operation(&mut self, operation: GeometricOperation) {
        if self.operations.contains_key(&operation.id) {
            return;
        }

        self.vector_clock.update(&operation.timestamp);
        self.operations.insert(operation.id, operation.clone());

        self.state = match operation.operation_type {
            OperationType::GeometricProduct => self.state.geometric_product(&operation.transform),
            OperationType::Addition => &self.state + &operation.transform,
            OperationType::Sandwich => {
                // R * v * R^-1 sandwich product
                let rev = operation.transform.reverse();
                operation
                    .transform
                    .geometric_product(&self.state)
                    .geometric_product(&rev)
            }
            OperationType::Exponential => operation.transform.exp().geometric_product(&self.state),
        };
    }

    /// Create a new operation to be applied
    pub fn create_operation(
        &mut self,
        transform: GA3,
        op_type: OperationType,
    ) -> GeometricOperation {
        self.vector_clock.tick(self.node_id);
        let op_id = self.operations.len() as u64;

        GeometricOperation {
            id: op_id,
            node_id: self.node_id,
            timestamp: self.vector_clock.clone(),
            transform,
            operation_type: op_type,
        }
    }

    /// Merge this CRDT with another, resolving conflicts using geometric algebra
    pub fn merge(&mut self, other: &GeometricCRDT) -> GeometricCRDT {
        let merged_clock = self.vector_clock.merge(&other.vector_clock);

        let mut merged_ops = self.operations.clone();
        for (id, op) in &other.operations {
            if !merged_ops.contains_key(id) {
                merged_ops.insert(*id, op.clone());
            }
        }

        // Re-apply all operations in causal order
        let mut sorted_ops: Vec<_> = merged_ops.values().cloned().collect();
        sorted_ops.sort_by(|a, b| {
            if a.timestamp.happens_before(&b.timestamp) {
                std::cmp::Ordering::Less
            } else if b.timestamp.happens_before(&a.timestamp) {
                std::cmp::Ordering::Greater
            } else {
                a.id.cmp(&b.id) // Deterministic tie-breaking
            }
        });

        let mut result = GeometricCRDT::new(self.node_id, GA3::zero());
        result.vector_clock = merged_clock;
        result.operations = merged_ops;

        for op in sorted_ops {
            result.apply_operation(op);
        }

        result
    }

    /// Compute geometric join for conflict resolution.
    ///
    /// Returns the state with larger magnitude, or their geometric mean if equal.
    pub fn geometric_join(&self, other: &GA3) -> GA3 {
        let self_norm = self.state.magnitude();
        let other_norm = other.magnitude();

        if self_norm > other_norm {
            self.state.clone()
        } else if other_norm > self_norm {
            other.clone()
        } else {
            // Equal magnitudes - use geometric mean
            geometric_mean(&[self.state.clone(), other.clone()])
        }
    }
}

/// Compute the geometric mean of a set of multivectors
pub fn geometric_mean(multivectors: &[GA3]) -> GA3 {
    if multivectors.is_empty() {
        return GA3::zero();
    }

    let n = multivectors.len() as f64;
    let sum_logs: GA3 = multivectors
        .iter()
        .map(|mv| mv.exp()) // Note: using exp as approximation since log may not exist for all
        .fold(GA3::zero(), |acc, log_mv| &acc + &log_mv);

    // Scale and return
    let coeffs: Vec<f64> = sum_logs.as_slice().iter().map(|&c| c / n).collect();
    GA3::from_slice(&coeffs)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_geometric_crdt_basic() {
        let node_id = Uuid::new_v4();
        let initial_state = GA3::scalar(1.0);
        let mut crdt = GeometricCRDT::new(node_id, initial_state);

        // Create and apply an addition operation
        let transform = GA3::scalar(2.0);
        let op = crdt.create_operation(transform, OperationType::Addition);
        crdt.apply_operation(op);

        // State should now be 3.0 (1.0 + 2.0)
        assert!((crdt.state.get(0) - 3.0).abs() < 1e-10);
    }

    #[test]
    fn test_geometric_crdt_convergence() {
        let node1_id = Uuid::new_v4();
        let node2_id = Uuid::new_v4();

        let initial_state = GA3::scalar(1.0);
        let mut crdt1 = GeometricCRDT::new(node1_id, initial_state.clone());
        let mut crdt2 = GeometricCRDT::new(node2_id, initial_state);

        // Node 1 applies an operation
        let op1 = crdt1.create_operation(GA3::scalar(2.0), OperationType::Addition);
        crdt1.apply_operation(op1.clone());

        // Node 2 applies a different operation
        let op2 = crdt2.create_operation(GA3::scalar(3.0), OperationType::Addition);
        crdt2.apply_operation(op2.clone());

        // Merge states - should converge
        let merged1 = crdt1.merge(&crdt2);
        let merged2 = crdt2.merge(&crdt1);

        // Both merges should produce the same state
        let diff = merged1.state.get(0) - merged2.state.get(0);
        assert!(diff.abs() < 1e-10);
    }

    #[test]
    fn test_vector_clock_ordering() {
        let mut clock1 = VectorClock::new();
        let mut clock2 = VectorClock::new();

        let node1 = Uuid::new_v4();
        let node2 = Uuid::new_v4();

        clock1.tick(node1);
        clock2.tick(node2);

        assert!(clock1.concurrent(&clock2));

        clock1.update(&clock2);
        clock1.tick(node1);

        assert!(clock2.happens_before(&clock1));
    }
}

File: cliffy-tsukoshi/src/protocols/crdt.ts
/**
 * Geometric CRDT (Conflict-free Replicated Data Type) using Clifford algebra.
 *
 * This CRDT uses geometric algebra operations for conflict resolution,
 * enabling coordination-free distributed state management.
 *
 * @example
 * ```typescript
 * import { GeometricCRDT, OperationType } from 'cliffy-tsukoshi/protocols';
 * import { scalar } from 'cliffy-tsukoshi';
 *
 * const nodeId = crypto.randomUUID();
 * const crdt = new GeometricCRDT(nodeId, scalar(0));
 *
 * // Apply a geometric transformation
 * const op = crdt.createOperation(scalar(5), OperationType.Addition);
 * crdt.applyOperation(op);
 *
 * console.log(crdt.state); // scalar(5)
 * ```
 */

import type { GA3 } from '../ga3.js';
import { add, zero, magnitude, geometricProduct, reverse, scale, fromCoefficients } from '../ga3.js';
import { VectorClock } from './vector-clock.js';

/**
 * Types of geometric operations supported by the CRDT.
 */
export enum OperationType {
  /** Geometric product: state * transform */
  GeometricProduct = 'geometric_product',
  /** Addition: state + transform */
  Addition = 'addition',
  /** Sandwich product: transform * state * reverse(transform) */
  Sandwich = 'sandwich',
  /** Exponential: exp(transform) * state */
  Exponential = 'exponential',
}

/**
 * A geometric operation that can be applied to the CRDT state.
 */
export interface GeometricOperation {
  /** Unique operation ID */
  id: number;
  /** Node that created this operation */
  nodeId: string;
  /** Timestamp for causal ordering */
  timestamp: VectorClock;
  /** The transformation to apply */
  transform: GA3;
  /** Type of operation */
  operationType: OperationType;
}

/**
 * A CRDT that uses geometric algebra operations for conflict resolution.
 */
export class GeometricCRDT {
  /** Current state as a multivector */
  state: GA3;
  /** Vector clock for causal ordering */
  vectorClock: VectorClock;
  /** This node's ID */
  nodeId: string;
  /** Applied operations (keyed by ID) */
  operations: Map<number, GeometricOperation>;
  /** Next operation ID */
  private nextOpId: number;

  constructor(nodeId: string, initialState: GA3) {
    this.state = initialState;
    this.vectorClock = new VectorClock();
    this.nodeId = nodeId;
    this.operations = new Map();
    this.nextOpId = 0;
  }

  /**
   * Apply a geometric operation to the CRDT state.
   */
  applyOperation(operation: GeometricOperation): void {
    // Skip if already applied (idempotent)
    if (this.operations.has(operation.id)) {
      return;
    }

    // Update vector clock
    this.vectorClock.update(operation.timestamp);

    // Store operation
    this.operations.set(operation.id, operation);

    // Apply transformation
    switch (operation.operationType) {
      case OperationType.GeometricProduct:
        this.state = geometricProduct(this.state, operation.transform);
        break;

      case OperationType.Addition:
        this.state = add(this.state, operation.transform);
        break;

      case OperationType.Sandwich:
        // R * v * R^-1 sandwich product
        const rev = reverse(operation.transform);
        const temp = geometricProduct(operation.transform, this.state);
        this.state = geometricProduct(temp, rev);
        break;

      case OperationType.Exponential:
        // exp(transform) * state
        const expTransform = gaExp(operation.transform);
        this.state = geometricProduct(expTransform, this.state);
        break;
    }
  }

  /**
   * Create a new operation to be applied.
   */
  createOperation(transform: GA3, opType: OperationType): GeometricOperation {
    this.vectorClock.tick(this.nodeId);
    const opId = this.nextOpId++;

    return {
      id: opId,
      nodeId: this.nodeId,
      timestamp: this.vectorClock.clone(),
      transform,
      operationType: opType,
    };
  }

  /**
   * Merge this CRDT with another, resolving conflicts using geometric algebra.
   */
  merge(other: GeometricCRDT): GeometricCRDT {
    const mergedClock = this.vectorClock.merge(other.vectorClock);

    // Collect all operations
    const mergedOps = new Map(this.operations);
    for (const [id, op] of other.operations) {
      if (!mergedOps.has(id)) {
        mergedOps.set(id, op);
      }
    }

    // Sort operations in causal order
    const sortedOps = Array.from(mergedOps.values()).sort((a, b) => {
      if (a.timestamp.happensBefore(b.timestamp)) {
        return -1;
      } else if (b.timestamp.happensBefore(a.timestamp)) {
        return 1;
      } else {
        // Deterministic tie-breaking by ID
        return a.id - b.id;
      }
    });

    // Re-apply all operations in order
    const result = new GeometricCRDT(this.nodeId, zero());
    result.vectorClock = mergedClock;
    result.operations = mergedOps;
    result.nextOpId = Math.max(this.nextOpId, other.nextOpId);

    for (const op of sortedOps) {
      result.applyOperation(op);
    }

    return result;
  }

  /**
   * Compute geometric join for conflict resolution.
   *
   * Returns the state with larger magnitude, or their geometric mean if equal.
   */
  geometricJoin(other: GA3): GA3 {
    const selfNorm = magnitude(this.state);
    const otherNorm = magnitude(other);

    if (selfNorm > otherNorm) {
      return this.state;
    } else if (otherNorm > selfNorm) {
      return other;
    } else {
      // Equal magnitudes - use geometric mean
      return geometricMean([this.state, other]);
    }
  }

  /**
   * Serialize to JSON.
   */
  toJSON(): object {
    return {
      state: this.state,
      vectorClock: this.vectorClock.toJSON(),
      nodeId: this.nodeId,
      operations: Array.from(this.operations.entries()).map(([id, op]) => ({
        id,
        nodeId: op.nodeId,
        timestamp: op.timestamp.toJSON(),
        transform: op.transform,
        operationType: op.operationType,
      })),
    };
  }

  /**
   * Deserialize from JSON.
   */
  static fromJSON(data: any): GeometricCRDT {
    const crdt = new GeometricCRDT(data.nodeId, data.state);
    crdt.vectorClock = VectorClock.fromJSON(data.vectorClock);
    for (const op of data.operations) {
      crdt.operations.set(op.id, {
        ...op,
        timestamp: VectorClock.fromJSON(op.timestamp),
      });
    }
    return crdt;
  }
}

/**
 * Compute the geometric mean of a set of multivectors.
 */
export function geometricMean(multivectors: GA3[]): GA3 {
  if (multivectors.length === 0) {
    return zero();
  }

  const n = multivectors.length;

  // Sum all (using exp as approximation since log may not exist for all)
  let sum = zero();
  for (const mv of multivectors) {
    sum = add(sum, gaExp(mv));
  }

  // Scale by 1/n
  return scale(sum, 1 / n);
}

/**
 * Simple exponential approximation for GA3.
 * Uses Taylor series: exp(x) ≈ 1 + x + x²/2! + x³/3! + ...
 */
function gaExp(mv: GA3): GA3 {
  const coeffs = mv as number[];
  const scalarPart = coeffs[0];

  // For primarily scalar values, use standard exp
  const vectorMag = Math.sqrt(
    coeffs.slice(1).reduce((sum, c) => sum + c * c, 0)
  );

  if (vectorMag < 1e-10) {
    // Pure scalar - standard exponential
    return fromCoefficients([Math.exp(scalarPart), 0, 0, 0, 0, 0, 0, 0]);
  }

  // General case: use Taylor series approximation (4 terms)
  let result = fromCoefficients([1, 0, 0, 0, 0, 0, 0, 0]); // 1
  let term = mv; // x
  result = add(result, term);

  term = scale(geometricProduct(term, mv), 0.5); // x²/2!
  result = add(result, term);

  term = scale(geometricProduct(term, mv), 1 / 3); // x³/3!
  result = add(result, term);

  term = scale(geometricProduct(term, mv), 0.25); // x⁴/4!
  result = add(result, term);

  return result;
}
